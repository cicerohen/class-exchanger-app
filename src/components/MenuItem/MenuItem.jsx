import React from 'react';
import PropTypes from 'prop-types';
import Style from './style/MenuItem.scss';

const MenuItem = ({
  children, href, title, onClick, style,
}) => (
  <li className={style.root}>
    <a className={style.container} href={href} title={title} onClick={onClick}>
      { children }
    </a>
  </li>
);

MenuItem.defaultProps = {
  href: '#',
  onClick: undefined,
  title: undefined,
  style: Style,
};

MenuItem.propTypes = {
  href: PropTypes.string,
  title: PropTypes.string,
  onClick: PropTypes.func,
  style: PropTypes.objectOf(PropTypes.any),
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
};

export default MenuItem;
