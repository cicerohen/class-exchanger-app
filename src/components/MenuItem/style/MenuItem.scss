@import './MenuItem.variables';

.root {
  list-style: none;
}

.container {
  text-decoration: none;
  padding-left: 15px;
  padding-top: 10px;
  padding-right: 15px;
  padding-bottom: 10px;
  display: block;
}
