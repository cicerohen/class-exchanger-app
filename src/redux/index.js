export * from './modules';
export * from './store';
