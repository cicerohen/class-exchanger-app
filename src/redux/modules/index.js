import * as ui from './ui';
import * as form from './form';
import * as firebase from './firebase';
import * as auth from './auth';
import * as router from './router';
import * as user from './user';

export {
  ui,
  form,
  firebase,
  auth,
  router,
  user,
};
