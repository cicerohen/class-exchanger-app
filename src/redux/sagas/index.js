import auth from './authSaga';
import firebase from './firebaseSaga';
import router from './routerSaga';
import user from './userSaga';
import exchange from './exchangeSaga';

export {
  auth,
  firebase,
  router,
  user,
  exchange,
};
