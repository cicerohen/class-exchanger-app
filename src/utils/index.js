import { Iterable, fromJS } from 'immutable';
import { createBrowserHistory } from 'history';

export const asImmutable = obj => (Iterable.isIterable(obj) ? obj : fromJS(obj));
export const emptyMap = asImmutable({});

export const buildActions = (moduleName, prefix, suffix) => ({
    REQUESTED: `${moduleName}/${prefix}_${suffix || ''}${suffix ? '_' : ''}REQUESTED`,
    STARTED: `${moduleName}/${prefix}_${suffix || ''}${suffix ? '_' : ''}STARTED`,
    SUCCEEDED: `${moduleName}/${prefix}_${suffix || ''}${suffix ? '_' : ''}SUCCEEDED`,
    FAILED: `${moduleName}/${prefix}_${suffix || ''}${suffix ? '_' : ''}FAILED`,
    FULFILLED: `${moduleName}/${prefix}_${suffix || ''}${suffix ? '_' : ''}FULFILLED`,
});

export const toArrayPath = path => path.split('/').filter(p => p.replace(' ', '').length > 0);

export const getHistory = () => {
    window.browserHistory = window.browserHistory || createBrowserHistory();
    return window.browserHistory;
};

export const isDevEnv = () => {
  const nodeEnv = process.env.NODE_ENV;
  return nodeEnv !== 'production' && nodeEnv !== 'test';
};
