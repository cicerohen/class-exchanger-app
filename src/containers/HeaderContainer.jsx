import React, { Component } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import PropTypes from 'prop-types';
import { auth, ui, router } from '../redux';
import Header from '../components/Header';

class HeaderContainer extends Component {
  openSidebar = () => {
    const { openSidebarAction } = this.props;
    openSidebarAction();
  }
  signOut = () => {
    const { signOutAction } = this.props;
    signOutAction();
  }

  push = (href) => {
    const { pushAction } = this.props;
    pushAction(href);
  }

  render() {
    const { isAuthenticated } = this.props;
    return (
      <Header
        openSidebarFn={this.openSidebar}
        isAuthenticated={isAuthenticated}
        signOutFn={this.signOut}
        pushFn={this.push}
      />
    );
  }
}

HeaderContainer.propTypes = {
  /** required */
  openSidebarAction: PropTypes.func.isRequired,
  pushAction: PropTypes.func.isRequired,
  signOutAction: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
};

const mapStateToProps = createStructuredSelector({
  isAuthenticated: auth.selectors.isAuthenticated,
});

const mapDispatchToProps = {
  openSidebarAction: ui.actions.openSidebar,
  pushAction: router.actions.push,
  signOutAction: auth.actions.signOut,
};

export default connect(mapStateToProps, mapDispatchToProps)(HeaderContainer);
