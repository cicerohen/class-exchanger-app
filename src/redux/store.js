import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { combineReducers } from 'redux-immutable';
import createSagaMiddleware from 'redux-saga';
import { createLogger } from 'redux-logger';
import { asImmutable, isDevEnv } from '../utils';
import * as modules from './modules';
import * as allSagas from './sagas';

const getReducers = () => Object.keys(modules).reduce((acc, curr) => {
    const module = modules[curr];
    if (module.reducers && module.reducers.reducer) {
        acc[module.constants.default] = module.reducers.reducer;
    }
    return acc;
}, {});

export const createReduxStore = ({
  middlewares = [],
  initialState = {},
  enhancers = [],
} = {}) => {
  const sagaMiddleware = createSagaMiddleware();
  if (isDevEnv()) {
    middlewares.push(createLogger({
      collapsed: true,
    }));
  }
  middlewares.push(sagaMiddleware);
  enhancers.push(applyMiddleware(...middlewares));
  const enhancer = composeWithDevTools(...enhancers);
  const store = createStore(
    combineReducers(getReducers()),
    asImmutable(initialState),
    enhancer,
  );
  store.sagas = Object.keys(allSagas).map(sagaName => sagaMiddleware.run(allSagas[sagaName]));
  store.cancelSagas = () => store.sagas.forEach(saga => saga.cancel());
  return store;
};
