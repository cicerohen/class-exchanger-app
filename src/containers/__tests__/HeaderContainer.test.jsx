import React from 'react';
import { shallow } from 'enzyme';
import { FaBars } from 'react-icons/lib/fa';
import MenuItem from '../../components/MenuItem';
import HeaderContainer from '../HeaderContainer';
import { createReduxTestStore } from '../../utils/test';
import { ui, auth, router } from '../../redux';

const eventMockObject = {
  preventDefault: jest.fn(),
};

describe('<HeaderContainer /> Tests', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render without exploding', (done) => {
    const testStore = createReduxTestStore();
    const wrapper = shallow(<HeaderContainer store={testStore} />);
    expect(wrapper).toHaveLength(1);
    done();
  });

  it('should render correctly when is not authenticated', (done) => {
    const testStore = createReduxTestStore({
      configs: {
        initialState: {
          auth: {
            isAuthenticated: false,
          },
        },
      },
    });
    const wrapper = shallow(<HeaderContainer store={testStore} />);
    expect(wrapper.dive().shallow()).toMatchSnapshot();
    done();
  });

  it('should render correctly when is authenticated', (done) => {
    const testStore = createReduxTestStore({
      configs: {
        initialState: {
          auth: {
            isAuthenticated: true,
          },
        },
      },
    });
    const wrapper = shallow(<HeaderContainer store={testStore} />);
    expect(wrapper.dive().shallow()).toMatchSnapshot();
    done();
  });

  it('should open sidebar when click on hamburguer menu item', (done) => {
    const testStore = createReduxTestStore({
      configs: {
        initialState: {
          ui: {
            sidebarIsOpen: false,
          },
        },
      },
    });
    const wrapper = shallow(<HeaderContainer store={testStore} />);
    testStore.when(ui.actions.OPEN_SIDEBAR, () => {
      const sidebarIsOpen = testStore.select(ui.selectors.sidebarIsOpen);
      expect(sidebarIsOpen).toBe(true);
      done();
    });

    const menuItem = wrapper.dive().shallow()
    .find(MenuItem)
    .filterWhere(item => item.children().type() === FaBars);

    const menuItemAnchor = menuItem.dive().shallow().find('a').first();
    menuItemAnchor.simulate('click', eventMockObject);
  });

  it('should sign out when click on sign out menu', (done) => {
    const testStore = createReduxTestStore({
      configs: {
        initialState: {
          auth: {
            isAuthenticated: true,
          },
        },
      },
    });
    const wrapper = shallow(<HeaderContainer store={testStore} />);
    testStore.when(auth.actions.SIGN_OUT.REQUESTED, () => {
      done();
    });
    const menuItem = wrapper.dive().shallow()
    .find(MenuItem)
    .filterWhere(item => item.children().contains('Sign Out'));
    const menuItemAnchor = menuItem.dive().shallow().find('a').first();
    menuItemAnchor.simulate('click', eventMockObject);
  });

  it('should navigate to Me page when click on menu item', (done) => {
    const testStore = createReduxTestStore({
      configs: {
        initialState: {
          auth: {
            isAuthenticated: true,
          },
        },
      },
    });
    const wrapper = shallow(<HeaderContainer store={testStore} />);
    testStore.when(router.actions.PUSH, () => {
      done();
    });
    const menuItem = wrapper.dive().shallow()
    .find(MenuItem)
    .filterWhere(item => item.children().contains('Me'));
    const menuItemAnchor = menuItem.dive().shallow().find('a').first();
    menuItemAnchor.simulate('click', eventMockObject);
  });

  it('should navigate to Home page when click on menu item', (done) => {
    const testStore = createReduxTestStore();
    const wrapper = shallow(<HeaderContainer store={testStore} />);
    testStore.when(router.actions.PUSH, () => {
      done();
    });
    const menuItem = wrapper.dive().shallow()
    .find(MenuItem)
    .filterWhere(item => item.children().contains('Home'));
    const menuItemAnchor = menuItem.dive().shallow().find('a').first();
    menuItemAnchor.simulate('click', eventMockObject);
  });

  it('should navigate to Sign In page when click on menu item', (done) => {
    const testStore = createReduxTestStore();
    const wrapper = shallow(<HeaderContainer store={testStore} />);
    testStore.when(router.actions.PUSH, () => {
      done();
    });
    const menuItem = wrapper.dive().shallow()
    .find(MenuItem)
    .filterWhere(item => item.children().contains('Sign In'));
    const menuItemAnchor = menuItem.dive().shallow().find('a').first();
    menuItemAnchor.simulate('click', eventMockObject);
  });

  it('should navigate to Sign Up page when click on menu item', (done) => {
    const testStore = createReduxTestStore();
    const wrapper = shallow(<HeaderContainer store={testStore} />);
    testStore.when(router.actions.PUSH, () => {
      done();
    });
    const menuItem = wrapper.dive().shallow()
    .find(MenuItem)
    .filterWhere(item => item.children().contains('Sign Up'));
    const menuItemAnchor = menuItem.dive().shallow().find('a').first();
    menuItemAnchor.simulate('click', eventMockObject);
  });
});
