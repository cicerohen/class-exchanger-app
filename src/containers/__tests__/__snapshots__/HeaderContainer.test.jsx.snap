// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderContainer /> Tests should render correctly when is authenticated 1`] = `
<header
  className="root"
>
  <div
    className="container"
  >
    <div
      className="contentLeftContainer"
    >
      <Menu
        style={
          Object {
            "root": "menu__root",
          }
        }
      >
        <MenuItem
          href="#"
          onClick={[Function]}
          style={
            Object {
              "container": "menuItem__container",
            }
          }
        >
          <FaBars />
        </MenuItem>
        <MenuItem
          href="/"
          onClick={[Function]}
          style={
            Object {
              "container": "menuItem__container",
            }
          }
        >
          Home
        </MenuItem>
      </Menu>
    </div>
    <div
      className="contentRightContainer"
    >
      <MenuItem
        href="me"
        onClick={[Function]}
        style={
          Object {
            "container": "menuItem__container",
          }
        }
      >
        Me
      </MenuItem>
      <MenuItem
        href="#"
        onClick={[Function]}
        style={
          Object {
            "container": "menuItem__container",
          }
        }
      >
        Sign Out
      </MenuItem>
    </div>
  </div>
</header>
`;

exports[`<HeaderContainer /> Tests should render correctly when is not authenticated 1`] = `
<header
  className="root"
>
  <div
    className="container"
  >
    <div
      className="contentLeftContainer"
    >
      <Menu
        style={
          Object {
            "root": "menu__root",
          }
        }
      >
        <MenuItem
          href="#"
          onClick={[Function]}
          style={
            Object {
              "container": "menuItem__container",
            }
          }
        >
          <FaBars />
        </MenuItem>
        <MenuItem
          href="/"
          onClick={[Function]}
          style={
            Object {
              "container": "menuItem__container",
            }
          }
        >
          Home
        </MenuItem>
      </Menu>
    </div>
    <div
      className="contentRightContainer"
    >
      <MenuItem
        href="signin"
        onClick={[Function]}
        style={
          Object {
            "container": "menuItem__container",
          }
        }
      >
        Sign In
      </MenuItem>
      <MenuItem
        href="signup"
        onClick={[Function]}
        style={
          Object {
            "container": "menuItem__container",
          }
        }
      >
        Sign Up
      </MenuItem>
    </div>
  </div>
</header>
`;
